src/main/scala/org/apache/spark/groupon/metrics/MetricsReceiver.scala

```diff
@@ -93,67 +91,94 @@ private[metrics] class MetricsReceiver(val sparkContext: SparkContext,
-    metrics.getOrElseUpdate(metricName, {
-      val histogram = new Histogram(reservoirClass.newInstance())
-      registerMetricSource(metricName, histogram)
-      histogram
-    }).asInstanceOf[Histogram]
+  def removeMetrics[T <: Metric](metricName: String, newMetric: => T): java.util.function.Function[String, Metric] = new function.Function[String, Metric] {
+    override def apply(t: String): Metric = {
+      removeMetricsSource(metricName, newMetric)
+      newMetric
+    }
   }
 
-  def getOrCreateMeter(metricName: String): Meter = {
-    metrics.getOrElseUpdate(metricName, {
-      val meter = new Meter()
-      registerMetricSource(metricName, meter)
-      meter
-    }).asInstanceOf[Meter]
+
+  def getOrCreateCounter(metricName: String): Counter =
+    metrics.computeIfAbsent(metricName, compute(metricName, new Counter)
+    ).asInstanceOf[Counter]
+
+
+  def getOrCreateHistogram(metricName: String, reservoirClass: Class[_ <: Reservoir]): Histogram =
+    metrics.computeIfAbsent(
+      metricName,
+      compute(metricName, new Histogram(reservoirClass.newInstance()))
+    ).asInstanceOf[Histogram]
+
+
+  def getOrCreateMeter(metricName: String): Meter =
+    metrics.computeIfAbsent(metricName, compute(metricName, new Meter)
+    ).asInstanceOf[Meter]
+
+
+  def getOrCreateTimer(metricName: String, reservoirClass: Class[_ <: Reservoir], clockClass: Class[_ <: Clock]): Timer =
+    metrics.computeIfAbsent(
+      metricName,
+      compute(metricName, new Timer(reservoirClass.newInstance(), clockClass.newInstance()))
+    ).asInstanceOf[Timer]
+
+  def getOrCreateGauge(metricName: String): Gauge[AnyVal] = {
+    metrics.computeIfAbsent(
+      metricName,
+      compute(metricName, new Gauge[AnyVal] {
+        override def getValue: AnyVal = lastGaugeValues.get(metricName)
+      })
+    ).asInstanceOf[Gauge[AnyVal]]
   }
 
-  def getOrCreateTimer(metricName: String, reservoirClass: Class[_ <: Reservoir], clockClass: Class[_ <: Clock]): Timer = {
-    metrics.getOrElseUpdate(metricName, {
-      val timer = new Timer(reservoirClass.newInstance(), clockClass.newInstance())
-      registerMetricSource(metricName, timer)
-      timer
-    }).asInstanceOf[Timer]
+  def remove(name: String, value: AnyVal): Unit = {
+    lastGaugeValues.remove(name)
+    metrics.remove(name)
+    metrics.computeIfAbsent(
+      name,
+      removeMetrics(name, new Gauge[AnyVal] {
+        override def getValue: AnyVal = lastGaugeValues.get(name)
+      })
+    ).asInstanceOf[Gauge[AnyVal]]
+    val metric = metrics.remove(name)
+    if (metric != null) return true
+    false
   }
 
-  def getOrCreateGauge(metricName: String): Gauge[AnyVal] = {
-    metrics.getOrElseUpdate(metricName, {
-      val gauge = new Gauge[AnyVal] {
-        override def getValue: AnyVal = {
-          lastGaugeValues.get(metricName).get
+  def removeMetricsSource(metricName: String, metric: Metric): Unit =  {
```

Comment:
When/where are these remove*() methods invoked?

Reply:
Using these remove method in SnappyMetricsSystem.scala class for deleting table metrics when the table gets deleted from cluster.